csharp/src/Apache.Arrow/Ipc/ArrowReaderImplementation.cs

```diff
@@ -0,0 +1,207 @@
+
+        private static bool MatchEnum(Flatbuf.MessageHeader messageHeader, Type flatBuffType)
+        {
+            switch (messageHeader)
+            {
+                case Flatbuf.MessageHeader.RecordBatch:
+                    return flatBuffType == typeof(Flatbuf.RecordBatch);
+                case Flatbuf.MessageHeader.DictionaryBatch:
+                    return flatBuffType == typeof(Flatbuf.DictionaryBatch);
+                case Flatbuf.MessageHeader.Schema:
+                    return flatBuffType == typeof(Flatbuf.Schema);
+                case Flatbuf.MessageHeader.Tensor:
+                    return flatBuffType == typeof(Flatbuf.Tensor);
+                case Flatbuf.MessageHeader.NONE:
+                    throw new ArgumentException("MessageHeader NONE has no matching flatbuf types", nameof(messageHeader));
+                default:
+                    throw new ArgumentException($"Unexpected MessageHeader value", nameof(messageHeader));
+            }
+        }
+
+        protected RecordBatch CreateArrowObjectFromMessage(Flatbuf.Message message, ByteBuffer bodyByteBuffer)
+        {
+            switch (message.HeaderType)
+            {
+                case Flatbuf.MessageHeader.Schema:
+                    // TODO: Read schema and verify equality?
+                    break;
+                case Flatbuf.MessageHeader.DictionaryBatch:
+                    // TODO: not supported currently
+                    Debug.WriteLine("Dictionaries are not yet supported.");
+                    break;
+                case Flatbuf.MessageHeader.RecordBatch:
+                    var rb = message.Header<Flatbuf.RecordBatch>().Value;
+                    var arrays = BuildArrays(Schema, bodyByteBuffer, rb);
+                    return new RecordBatch(Schema, arrays, (int)rb.Length);
+                default:
+                    // NOTE: Skip unsupported message type
+                    Debug.WriteLine($"Skipping unsupported message type '{message.HeaderType}'");
+                    break;
+            }
+
+            return null;
+        }
+
+        private IEnumerable<IArrowArray> BuildArrays(
+            Schema schema,
+            ByteBuffer messageBuffer,
+            Flatbuf.RecordBatch recordBatchMessage)
+        {
+            var arrays = new List<ArrayData>();
+            var bufferIndex = 0;
+
+            for (var n = 0; n < recordBatchMessage.NodesLength; n++)
+            {
+                var field = schema.GetFieldByIndex(n);
+                var fieldNode = recordBatchMessage.Nodes(n).GetValueOrDefault();
+
+                if (field.DataType.IsFixedPrimitive())
+                    arrays.Add(LoadPrimitiveField(field, fieldNode, recordBatchMessage, messageBuffer, ref bufferIndex));
+                else
+                    arrays.Add(LoadVariableField(field, fieldNode, recordBatchMessage, messageBuffer, ref bufferIndex));
+            }
+
+            return arrays.Select(ArrowArrayFactory.BuildArray);
```

Comment:
C# doesn't yet cache delegates for method groups, unfortunately.

Reply:
Will fix.